gninasrc/lib/device_buffer.cpp

```diff
@@ -7,48 +7,59 @@
 #include "gpu_util.h"
 #include <cassert>
 
-template<typename T>
-device_buffer<T>::device_buffer(size_t capacity) : capacity(capacity){
-    CUDA_CHECK_GNINA(cudaMalloc(&begin, capacity*sizeof(T)));
-    current = begin;
-}
-
-template<typename T>
-bool device_buffer<T>::has_space(size_t n_requested) {
-    size_t space_used = current - begin; //in elements
-    return capacity - space_used >= n_requested;
-}
-
-template<typename T>
-void device_buffer<T>::resize(size_t n_requested) {
-    if (begin == current && capacity >= n_requested)
-        return;
-    //This doesn't memcpy the old contents because of the baked-in assumption
-    //that you are resizing before you start copying. 
-    else if (!has_space(n_requested)) {
-        int factor = n_requested % capacity ? ((n_requested/capacity)+1)*capacity: n_requested;
-        size_t new_capacity = capacity * factor;
+#define align_down_pow2(n, size)            \
+    ((decltype (n)) ((uintptr_t) (n) & ~((size) - 1)))
+
+#define align_up_pow2(n, size)                                    \
+    ((decltype (n)) align_down_pow2((uintptr_t) (n) + (size) - 1, size))
+
+thread_local device_buffer thread_buffer(32 * 4096);
```

Comment:
Just making sure - the constructor argument of device_buffer is the initial capacity and it will resize as necessary, right?